File: Python/algorithmic_heights/15_bip.py
# Testing Bipartiteness
import sys

from collections import defaultdict, deque
from enum import Enum, auto


class NodeColor(Enum):
    RED = auto()
    BLUE = auto()


class Graph:
    def __init__(self, nr_nodes, edge_list):
        self._nr_nodes = nr_nodes
        self._adjacency_list = self._build_adjacency_list(edge_list)

    @staticmethod
    def _build_adjacency_list(edge_list):
        adjacency_list = defaultdict(list)
        for a, b in edge_list:
            adjacency_list[a].append(b)
            adjacency_list[b].append(a)
        return dict(adjacency_list)

    @property
    def nr_nodes(self):
        return self._nr_nodes

    def neighbours(self, node):
        return self._adjacency_list.get(node, [])


def convert_to_intlist(line):
    return tuple(int(item) for item in line.split())


def read_edges(reader, nr_edges):
    edges = []
    for _ in range(nr_edges):
        a, b = convert_to_intlist(next(reader))
        edges.append((a, b))
    return edges


def read_graphs(reader, n):
    graphs = []
    for _ in range(n):
        _ = next(reader)
        nr_nodes, nr_edges = convert_to_intlist(next(reader))
        edge_list = read_edges(reader, nr_edges)
        graph = Graph(nr_nodes, edge_list)
        graphs.append(graph)
    return graphs


def get_opposite_color(color):
    if color == NodeColor.RED:
        return NodeColor.BLUE

    return NodeColor.RED


def find_consistent_component_coloring(graph, starting_node):
    node_colors_in_component = {starting_node: NodeColor.RED}
    queue = deque([starting_node])
    while queue:
        node = queue.popleft()
        node_color = node_colors_in_component[node]
        color = get_opposite_color(node_color)
        neighbours = graph.neighbours(node)
        for neighbour in neighbours:
            if neighbour in node_colors_in_component:
                if node_colors_in_component[neighbour] == node_color:
                    return None
            else:
                node_colors_in_component[neighbour] = color
                queue.append(neighbour)
    return node_colors_in_component


def is_bipartite(graph):
    node_colors = {}
    for starting_node in range(1, graph.nr_nodes+1):
        if starting_node not in node_colors:
            component_coloring = find_consistent_component_coloring(graph, starting_node)
            if component_coloring is None:
                return False

            node_colors.update(component_coloring)
    return True


def test_bipartiteness(graphs):
    return [is_bipartite(graph) for graph in graphs]


def main():
    reader = sys.stdin
    n = int(next(reader))
    graphs = read_graphs(reader, n)
    result = test_bipartiteness(graphs)
    print(' '.join(map(lambda x: '1' if x else '-1', result)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/15_bip.hs
{-# LANGUAGE TupleSections #-}

import Control.Monad (mapM, replicateM)
import Data.IntMap (IntMap, (!))
import qualified Data.IntMap as M
import Data.List (partition)
import Data.Sequence ((><), ViewL ((:< ), EmptyL))
import qualified Data.Sequence as S

type AdjacencyList = IntMap [Int]

data NodeColor = Red | Blue deriving Eq
data Edge = Edge Int Int
data Graph = Graph { nrNodes :: Int, adjacencyList :: AdjacencyList }


convertToIntList :: String -> [Int]
convertToIntList = map read . words


getOppositeColor :: NodeColor -> NodeColor
getOppositeColor Red = Blue
getOppositeColor Blue = Red


readInputLine :: String -> (Int, Int)
readInputLine line =
    case convertToIntList line of
        [a, b] -> (a, b)
        _ -> error "Malformed input."


createGraph :: Int -> [Edge] -> Graph
createGraph nrNodes edges =
    let addEdge adjacencyList (Edge a b) =
            let update Nothing = Just [b]
                update (Just ns) = Just (b : ns)
            in M.alter update a adjacencyList
        addEdgeBothWays adjacencyList edge @ (Edge a b) =
            addEdge (addEdge adjacencyList edge) (Edge b a)
        adjacencyList = foldl addEdgeBothWays M.empty edges
    in Graph { nrNodes = nrNodes, adjacencyList = adjacencyList }


readGraphs :: Int -> IO [Graph]
readGraphs n =
    let readEdges nrEdges = do
            edges <- replicateM nrEdges getLine
            return $ map (\line -> let (a, b) = readInputLine line in Edge a b) edges
        readGraph = do
            _ <- getLine
            (nrNodes, nrEdges) <- readInputLine <$> getLine
            edges <- readEdges nrEdges
            return $ createGraph nrNodes edges
    in mapM (const readGraph) [1..n]


getNeighbors :: Graph -> Int -> [Int]
getNeighbors (Graph _ adjacencyList) node = M.findWithDefault [] node adjacencyList


findConsistentComponentColoring :: Graph -> Int -> Maybe (IntMap NodeColor)
findConsistentComponentColoring graph startNode =
    let go queue componentColors =
            case S.viewl queue of
                    EmptyL -> Just componentColors
                    (node :< rest) ->
                        let color = componentColors ! node
                            oppositeColor = getOppositeColor color
                            neighbors = getNeighbors graph node
                            (coloredNeighbors, uncoloredNeighbors) = partition (`M.member` componentColors) neighbors
                        in if any (\n -> componentColors ! n == color) coloredNeighbors then Nothing
                            else
                                let newNodes = M.fromList $ map (, oppositeColor) uncoloredNeighbors
                                in go (rest >< S.fromList uncoloredNeighbors) (M.union componentColors newNodes)
    in go (S.singleton startNode) (M.singleton startNode Red)


isBipartite :: Graph -> Bool
isBipartite graph =
    let go nodeColors startNode
            | startNode >= nrNodes graph = True
            | M.member startNode nodeColors = go nodeColors (startNode + 1)
            | otherwise =
                case findConsistentComponentColoring graph startNode of
                    Nothing -> False
                    Just componentColoring -> go (M.union nodeColors componentColoring) (startNode + 1)
    in go M.empty 1


main :: IO ()
main = do
    n <- readLn
    graphs <- readGraphs n
    let results = map isBipartite graphs
    putStrLn $ unwords $ map (\res -> if res then "1" else "-1") results
